### PR TITLE
8288783: Error messages are confusing when options conflict in -XX:StartFlightRecording

### PR DESCRIPTION
**Pull request by chiroito**
Could I have a review of PR that fixes incorrect error messages regarding of recording.

The current message is very confusing. If a user types jcmd <pid> JFR.start name=abc name=def or java --XX:StartFlightRecording=name=abc,name=def,  the error message says "Duplicates in diagnostic command arguments", but doesn't know what options are duplicated.
Furthermore, if a user specifies two or three duplicate parameters, the same logs will be output.

It should say follows,
Option name can only be specified once with starting flight recording
Options name and disk can only be specified once with starting flight recording
Options name, disk and maxage can only be specified once with starting flight recording
This problem affects --XX:StartFlightRecording, jcmd JFR.start, JFR.stop, JFR.dump, JFR.check, and MBean.

For other than start, the output is as follows, (example of dump)
Option name can only be specified once with dumping flight recording

Testing: jdk/jdk/jfr

Thanks
Chihiro

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288783](https://bugs.openjdk.org/browse/JDK-8288783): Error messages are confusing when options conflict in -XX:StartFlightRecording

### Reviewers
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9302/head:pull/9302` \
`$ git checkout pull/9302`

Update a local copy of the PR: \
`$ git checkout pull/9302` \
`$ git pull https://git.openjdk.org/jdk pull/9302/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9302`

View PR using the GUI difftool: \
`$ git pr show -t 9302`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9302.diff">https://git.openjdk.org/jdk/pull/9302.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jfr/share/classes/jdk/jfr/internal/dcmd/ArgumentParser.java [hunk 1]
```diff
@@ -64,10 +69,44 @@ public Map<String, Object> parse() {
             addOption(key, value);
             eatDelimiter();
         }
+        checkConflict();
         checkMandatory();
         return options;
     }
 
+    protected void checkConflict() {
```
Comment (egahlin) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/dcmd/ArgumentParser.java:77]: I think this part can be simplified, the variable names are long and there is too much state to track.

How about this?

`private void checkConflicted() {
        if (conflictedOptions.isEmpty()) {
            return;
        }

        StringBuilder sb = new StringBuilder();
        if (conflictedOptions.size() == 1) {
            sb.append("Option ");
            sb.append(conflictedOptions.remove(0));
        } else {
            sb.append("Options ");
            StringJoiner sj = new StringJoiner(", ");
            while (conflictedOptions.size() > 1) {
                sj.add(conflictedOptions.remove(0));
            }
            sb.append(sj);
            sb.append(" and " + conflictedOptions.remove(0));
        }
        sb.append(" can only be specified once.");
        throw new IllegalArgumentException(sb.toString());
    }`
    
It requires an ArrayList, but you could do:
`    if (!conflictedOptions.contains(key)) {
        conflictedOptions.add(key);
      }`
          
I don't think it's worth the complication with the JfrCommand enum to display the correct command being used in an error message.

Comment (chiroito) [reply src/jdk.jfr/share/classes/jdk/jfr/internal/dcmd/ArgumentParser.java:77]: Thank you for your prompt reviewing.

Your idea is good for me. I will change the code in that regard as soon as I can.

I agree with the complication you mentiond. We posted about this in JBS with phrases like "with -XX:StartFlightRecording" at the end of the log message, and also mentioned jcmd. Howeber, at the existing implementation in logging the message, there is no way to identify a command such as "JFR.start". So I implemented it this way to identify the command as a last resort. 

As you say, I also think it no need to have a log message that identifies each command with the complication. Therefore, I will eliminate this complication.

**Comment by chiroito** [comment]
@egahlin 
Could you review this again, please?

File: src/jdk.jfr/share/classes/jdk/jfr/internal/dcmd/ArgumentParser.java [hunk 2]
```diff
@@ -94,9 +123,12 @@ private void addOption(String key, String value) {
                     }
                 } else {
                     if (options.containsKey(key)) {
-                        throw new IllegalArgumentException("Duplicates in diagnostic command arguments");
+                        if(!conflictedOptions.contains(key)) {
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/dcmd/ArgumentParser.java:126]:
```suggestion
                        if (!conflictedOptions.contains(key)) {
```

File: test/jdk/jdk/jfr/jcmd/TestJcmdOptionSpecifiedOnce.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    private static void testMultipleOption(String command, String option){
+
+        String output = JcmdHelper.jcmd("%s %s".formatted(command, option)).getOutput();
+        final String regex = "Option ([a-z-]+) can only be specified once with ([a-z]+) flight recording";
+
+        try {
+            Matcher matcher = Pattern.compile(regex).matcher(output);
+            if(matcher.find()){
```
Comment (turbanoff) [inline test/jdk/jdk/jfr/jcmd/TestJcmdOptionSpecifiedOnce.java:123]:
```suggestion
            if (matcher.find()){
```

File: test/jdk/jdk/jfr/startupargs/TestStartupOptionSpecifiedOnce.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    private static void testMultipleOption(String option) throws Exception{
+
+        String output = startJfrJvm(option).getOutput();
+        final String regex = "Option ([a-z-]+) can only be specified once";
+
+        try {
+            Matcher matcher = Pattern.compile(regex).matcher(output);
+            if(matcher.find()){
```
Comment (turbanoff) [inline test/jdk/jdk/jfr/startupargs/TestStartupOptionSpecifiedOnce.java:101]:
```suggestion
            if (matcher.find()){
```

**Comment by egahlin** [comment]
What worries me a bit is the testing. It is thorough, but perhaps too thorough? 

It might fail due to other reasons than the added logic, or it may become a maintenance burden. For example, the flush-interval option is planned to be removed. Starting 40 JVMs to verify an improved error messages seems excessive. 

I would put the tests in one file, TestConflictingOptions and use options that are harmless (disk and name), but include "settings" as it should be repeatable, i.e 

    void testJCmdConflict() {
      var oa = JcmdHelper.jcmd("JFR.start name=hello name=greetings");
      oa.shouldContain("name can only be specified once")
    }

    void testStartFlightRecordingConflict() {
     var output = ProcessTools.executeTestJava("-XX:StartFlightRecording:disk=true,disk=false,name=cat,name=dog");
     output.shouldContain("disk and name can only be specified once.")
    }

    void testSettings() {
     var output = ProcessTools.executeTestJava("-XX:StartFlightRecording:settings=default,settings=profile");
     output.shouldNotContain("setting can only be specified once")
   }

This should give 100% line coverage and make sure it works from both command line and jcmd.

**Review by egahlin — APPROVED** [review]

**Comment by chiroito** [comment]
@egahlin Thanks for the review.

/integrate

**Pull request closed without merging**